### ISSUE TITLE
Can't insert data into three related tables

### ISSUE THREAD

**Issue by ranzer** [city41/node-sql-fixtures#31]
I have following tables in MariaDB v10.0.16 database:

A(
   id tinyint unsigned auto_increment primary key,
   title varchar(30) not null,
   unique(title)
),
B(
   id smallint unsigned auto_increment primary key,
   title varchar(30) not null,
   description varchar(500) not null,
   created datetime not null,
   a_id tinyint unsigned not null,
   unique(title),
   constraint B_A_a_id_id
   foreign key (a_id)
   references A (id)
   on delete cascade
   on update restrict
),
C(
  id smallint unsigned not null,
  b_id smallint unsigned not null,
  data varchar(1000) not null,
  primary key (id, b_id),
  constraint C_B_b_id_id
  foreign key (b_id)
  references B (id)
  on delete cascade
  on update restrict
)

When I try to insert following data: 

dataSpec = {
        A: [
          { title: 'A1' },
          { title: 'A2' },
          { title: 'A3' }
        ],
        B: [
          { title: 'B1', description: 'D1', created: new Date(), a_id: 'A:0' },
          { title: 'B2', description: 'D2', created: new Date(), a_id: 'A:0' },
          { title: 'B3', description: 'D3', created: new Date(), a_id: 'A:1' },
          { title: 'B4', description: 'D4', created: new Date(), a_id: 'A:1' },
          { title: 'B5', description: 'D5', created: new Date(), a_id: 'A:2' }
        ],
        C: [
          { id: 1, b_id: 'B:0', data: 'Test 1' },
          { id: 2, b_id: 'B:0', data: 'Test 2' },
          { id: 3, b_id: 'B:0', data: 'Test 3' },
          { id: 4, b_id: 'B:0', data: 'Test 4' },
          { id: 5, b_id: 'B:0', data: 'Test 5' }
        ]
      }
,
I get the error: 

Unhandled rejection Error: ER_NO_REFERENCED_ROW_2: Cannot add or update a child row: a foreign key constraint fails (`some_database`.`C`, CONSTRAINT `C_B_b_id_id` FOREIGN KEY (`b_id`) REFERENCES `B` (`id`) ON DELETE CASCADE)

When I remove the "created" column in table "B" everything works fine except that the program doesn't exit. 
If I just remove data related to the "C" table, from the dataSpec object, without removing the "created" column", the data is inserted, into tables A and B, as expected without an error and the program exits fine. 
I'm using sql-fixtures version 0.10.1. 
Why is this happening? 
Is this an issue or expected behaviour?

**Comment by city41** [comment]
Could you add `debug: true` to your knex config and include its output here? It should be a bunch of statements like this:

```
{ __cid: '__cid2',
  method: 'insert',
  options: undefined,
  bindings: [ 'a value' ],
  sql: 'insert into "simple_table" ("string_column") values (?)' }
```

```
var myKnex = knex({
    debug: true,
    // all the rest
});
```

**Comment by ranzer** [comment]
The output after adding the "debug" property:

{ __cid: '__cid1',
  method: 'insert',
  options: undefined,
  bindings: [ 'A1' ],
  sql: 'insert into `A` (`title`) values (?)' }
{ __cid: '__cid2',
  method: 'select',
  options: undefined,
  bindings: [ 'A1', 1 ],
  sql: 'select * from `A` where `title` = ? order by `id` DESC limit ?' }
{ __cid: '__cid3',
  method: 'insert',
  options: undefined,
  bindings: [ 'A2' ],
  sql: 'insert into `A` (`title`) values (?)' }
{ __cid: '__cid1',
  method: 'select',
  options: undefined,
  bindings: [ 'A2', 1 ],
  sql: 'select * from `A` where `title` = ? order by `id` DESC limit ?' }
{ __cid: '__cid2',
  method: 'insert',
  options: undefined,
  bindings: [ 'A3' ],
  sql: 'insert into `A` (`title`) values (?)' }
{ __cid: '__cid3',
  method: 'select',
  options: undefined,
  bindings: [ 'A3', 1 ],
  sql: 'select * from `A` where `title` = ? order by `id` DESC limit ?' }
{ __cid: '__cid1',
  method: 'insert',
  options: undefined,
  bindings: [ 4, Sun May 31 2015 18:54:24 GMT+0200 (CEST), 'D1', 'B1' ],
  sql: 'insert into `B` (`a_id`, `created`, `description`, `title`) values (?, ?, ?, ?)' }
{ __cid: '__cid2',
  method: 'select',
  options: undefined,
  bindings: [ 'B1', 'D1', Sun May 31 2015 18:54:24 GMT+0200 (CEST), 4, 1 ],
  sql: 'select * from `B` where `title` = ? and `description` = ? and `created` = ? and `a_id` = ? order by `id` DESC limit ?' }
{ __cid: '__cid3',
  method: 'insert',
  options: undefined,
  bindings: [ 4, Sun May 31 2015 18:54:24 GMT+0200 (CEST), 'D2', 'B2' ],
  sql: 'insert into `B` (`a_id`, `created`, `description`, `title`) values (?, ?, ?, ?)' }
{ __cid: '__cid1',
  method: 'select',
  options: undefined,
  bindings: [ 'B2', 'D2', Sun May 31 2015 18:54:24 GMT+0200 (CEST), 4, 1 ],
  sql: 'select * from `B` where `title` = ? and `description` = ? and `created` = ? and `a_id` = ? order by `id` DESC limit ?' }
{ __cid: '__cid2',
  method: 'insert',
  options: undefined,
  bindings: [ 5, Sun May 31 2015 18:54:24 GMT+0200 (CEST), 'D3', 'B3' ],
  sql: 'insert into `B` (`a_id`, `created`, `description`, `title`) values (?, ?, ?, ?)' }
{ __cid: '__cid3',
  method: 'select',
  options: undefined,
  bindings: [ 'B3', 'D3', Sun May 31 2015 18:54:24 GMT+0200 (CEST), 5, 1 ],
  sql: 'select * from `B` where `title` = ? and `description` = ? and `created` = ? and `a_id` = ? order by `id` DESC limit ?' }
{ __cid: '__cid1',
  method: 'insert',
  options: undefined,
  bindings: [ 5, Sun May 31 2015 18:54:24 GMT+0200 (CEST), 'D4', 'B4' ],
  sql: 'insert into `B` (`a_id`, `created`, `description`, `title`) values (?, ?, ?, ?)' }
{ __cid: '__cid2',
  method: 'select',
  options: undefined,
  bindings: [ 'B4', 'D4', Sun May 31 2015 18:54:24 GMT+0200 (CEST), 5, 1 ],
  sql: 'select * from `B` where `title` = ? and `description` = ? and `created` = ? and `a_id` = ? order by `id` DESC limit ?' }
{ __cid: '__cid3',
  method: 'insert',
  options: undefined,
  bindings: [ 6, Sun May 31 2015 18:54:24 GMT+0200 (CEST), 'D5', 'B5' ],
  sql: 'insert into `B` (`a_id`, `created`, `description`, `title`) values (?, ?, ?, ?)' }
{ __cid: '__cid1',
  method: 'select',
  options: undefined,
  bindings: [ 'B5', 'D5', Sun May 31 2015 18:54:24 GMT+0200 (CEST), 6, 1 ],
  sql: 'select * from `B` where `title` = ? and `description` = ? and `created` = ? and `a_id` = ? order by `id` DESC limit ?' }
{ __cid: '__cid2',
  method: 'insert',
  options: undefined,
  bindings: [ 'B:__genned_specId_3', 'Test 1', 1 ],
  sql: 'insert into `C` (`b_id`, `data`, `id`) values (?, ?, ?)' }

**Comment by city41** [comment]
Thanks. I've been able to reproduce the issue and I've created a failing test case for it. Your situation hits upon sql-fixture's achilles heal in a new way. I need to think about this and see if I can make that part of the code better once and for all.

**Comment by city41** [comment]
Removing the created column (or at least not populating it) will work around your issue, although you probably don't want to do that. I can't find any reason the program shouldn't end. The only thing I can think of is you need to call `sqlFixtures.destroy()` at some point, otherwise internally Knex will maintain a connection to your database, preventing the app from closing. With mocha, I usually do:

```
after(function(done) {
    sqlFixtures.destroy(done);
});
```

**Comment by city41** [comment]
This is now fixed in `0.10.2`, I added [this spec](https://github.com/city41/node-sql-fixtures/blob/master/test/integration/maria-integration-spec.js#L24) which is your situation exactly, so feeling confident about the fix.

MySQL/Maria still have some situations where sql-fixtures can fail, thankfully those situations are getting more and more obscure. Not sure I'll ever get 100% perfect MySql/Maria support. I now know how to get 100% perfect support for sqlite, which I have as a todo.
